Microsoft.Dotnet.Compatibility analyzer doesn't work with the NumericIntPtr feature

https://github.com/dotnet/runtime/pull/69627 updated the dotnet/runtime repo to utilize the `NumericIntPtr` feature from C# 11: https://github.com/dotnet/csharplang/issues/6065

This resulted in the compatibility analyzer misfiring with many errors such as:
```
/__w/1/s/.packages/microsoft.dotnet.compatibility/2.0.0-alpha.1.21525.11/build/Microsoft.NET.Compatibility.Common.targets(33,5): error CP0002: Member 'Microsoft.Extensions.Caching.Memory.PostEvictionDelegate.PostEvictionDelegate(object, System.IntPtr)' exists on lib/net6.0/Microsoft.Extensions.Caching.Abstractions.dll but not on lib/net7.0/Microsoft.Extensions.Caching.Abstractions.dll [/__w/1/s/src/libraries/Microsoft.Extensions.Caching.Abstractions/src/Microsoft.Extensions.Caching.Abstractions.csproj]
##[error].packages/microsoft.dotnet.compatibility/2.0.0-alpha.1.21525.11/build/Microsoft.NET.Compatibility.Common.targets(33,5): error CP0002: (NETCORE_ENGINEERING_TELEMETRY=Build) Member 'Microsoft.Extensions.Caching.Memory.PostEvictionDelegate.PostEvictionDelegate(object, System.IntPtr)' exists on lib/net6.0/Microsoft.Extensions.Caching.Abstractions.dll but not on lib/net7.0/Microsoft.Extensions.Caching.Abstractions.dll
```

This is despite the underlying assemblies still containing the correct signatures. I noticed the version was out of date and so grabbed the latest package available, 2.0.0-preview.4.22252.4, to see the same error.

--- a/src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Extensions/SymbolExtensions.cs
+++ b/src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Extensions/SymbolExtensions.cs
@@ -29,7 +29,10 @@ namespace Microsoft.DotNet.ApiCompatibility.Extensions
             return format.WithParameterOptions(format.ParameterOptions & ~SymbolDisplayParameterOptions.IncludeParamsRefOut);
         }
 
-        internal static string ToComparisonDisplayString(this ISymbol symbol) => symbol.ToDisplayString(Format);//.Replace("System.IntPtr", "nint");
+        internal static string ToComparisonDisplayString(this ISymbol symbol) =>
+            symbol.ToDisplayString(Format)
+                  .Replace("System.IntPtr", "nint") // Treat IntPtr and nint as the same
+                  .Replace("System.UIntPtr", "nuint"); // Treat UIntPtr and nuint as the same
 
         internal static IEnumerable<ITypeSymbol> GetAllBaseTypes(this ITypeSymbol type)
         {

--- a/src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Extensions/SymbolExtensions.cs
+++ b/src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Extensions/SymbolExtensions.cs
@@ -29,7 +29,7 @@ namespace Microsoft.DotNet.ApiCompatibility.Extensions
             return format.WithParameterOptions(format.ParameterOptions & ~SymbolDisplayParameterOptions.IncludeParamsRefOut);
         }
 
-        internal static string ToComparisonDisplayString(this ISymbol symbol) => symbol.ToDisplayString(Format);
+        internal static string ToComparisonDisplayString(this ISymbol symbol) => symbol.ToDisplayString(Format);//.Replace("System.IntPtr", "nint");
 
         internal static IEnumerable<ITypeSymbol> GetAllBaseTypes(this ITypeSymbol type)
         {

--- a/src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs
+++ b/src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs
@@ -466,5 +466,41 @@ namespace CompatTests
             };
             Assert.Equal(expected, differences);
         }
+
+        [Fact]
+        public void NumericPtrFlaggedOnlyBeforeNet7()
+        {
+            string leftSyntax = @"
+namespace CompatTests
+{
+  using System;
+
+  public class First
+  {
+    public void F(IntPtr p) {}
+  }
+}
+";
+            string rightSyntax = @"
+namespace CompatTests
+{
+  public class First
+  {
+    public void F(nint p) {}
+  }
+}
+";
+            IAssemblySymbol left = SymbolFactory.GetAssemblyFromSyntax(leftSyntax);
+            IAssemblySymbol right = SymbolFactory.GetAssemblyFromSyntax(rightSyntax);
+            ApiComparer differ = new(s_ruleFactory);
+
+            IEnumerable<CompatDifference> differences = differ.GetDifferences(left, right);
+
+            CompatDifference[] expected =
+            {
+            };
+            // ReferenceID on .NET 7 for left member is "M:CompatTests.First.F(System.IntPtr)"
+            Assert.Equal(expected, differences);
+        }
     }
 }

--- a/src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs
+++ b/src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs
@@ -468,7 +468,7 @@ namespace CompatTests
         }
 
         [Fact]
-        public void NumericPtrFlaggedOnlyBeforeNet7()
+        public void NumericPtrNotFlagged()
         {
             string leftSyntax = @"
 namespace CompatTests
@@ -478,6 +478,7 @@ namespace CompatTests
   public class First
   {
     public void F(IntPtr p) {}
+    public void G(UIntPtr p) {}
   }
 }
 ";
@@ -487,6 +488,7 @@ namespace CompatTests
   public class First
   {
     public void F(nint p) {}
+    public void G(nuint p) {}
   }
 }
 ";
@@ -496,11 +498,7 @@ namespace CompatTests
 
             IEnumerable<CompatDifference> differences = differ.GetDifferences(left, right);
 
-            CompatDifference[] expected =
-            {
-            };
-            // ReferenceID on .NET 7 for left member is "M:CompatTests.First.F(System.IntPtr)"
-            Assert.Equal(expected, differences);
+            Assert.Empty(differences);
         }
     }
 }

